[SelectionDAG] Add an ISD::CLEAR_CACHE node to lower llvm.clear_cache

rofirrim opened the pull request:
The current way of lowering `llvm.clear_cache` is a bit unusual. As suggested by @arsenm we are better off using an ISD node.

This change introduces a new `ISD::CLEAR_CACHE`, registers a new libcall by default named `__clear_cache` and the default legalisation is a libcall.

This is preparatory work for a custom lowering of `ISD::CLEAR_CACHE` needed by RISC-V on some platforms.

rofirrim committed:
[SelectionDAG] Add ISD::CLEAR_CACHE

The existing way to lower llvm.clear_cache is not ideal. As suggested by
Matt Arsenault, let's turn this into a regular ISD node.

This change adds the node itself and nothing else.

rofirrim committed:
[SelectionDAG] Plumbing lowering of ISD::CLEAR_CACHE as a libcall in TLI

Lower the newly introduced ISD::CLEAR_CACHE as a runtime function. By
default we lower it to __clear_cache as customary.

rofirrim committed:
[X86][Webassembly] Update targets that need non-default clear_cache lowering

Some targets do not lower llvm.clear_cache to a call to __clear_cache.
Update those to handle the new ISD::CLEAR_CACHE.

rofirrim committed:
[SelectionDAG] Remove old getClearCacheBuiltinName hook

Now we can remove the old hook.

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -32387,6 +32390,10 @@ SDValue X86TargetLowering::LowerOperation(SDValue Op, SelectionDAG &DAG) const {
   case X86ISD::CVTPS2PH:        return LowerCVTPS2PH(Op, DAG);
   case ISD::PREFETCH:           return LowerPREFETCH(Op, Subtarget, DAG);
+  case ISD::CLEAR_CACHE: {
+    // X86 instruction cache is coherent with its data cache so this is a no-op.
+    return Op->getOperand(0);
```
Just delete the operation should probably be the default expansion. Probably should move this comment to the setOperationAction, and move this implementation to the default expand action

rofirrim committed:
[SelectionDAG] Make the expansion of ISD::CLEAR_CACHE be a no-op

Some targets like X86 have coherent instruction and data caches
so this expansion is convenient to have.

rofirrim committed:
clang-format

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7516,11 +7516,18 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
-      lowerCallToExternalSymbol(I, FunctionName);
+  case Intrinsic::clear_cache: {
+    SDValue InputChain = DAG.getRoot();
+    Value *Start = I.getArgOperand(0);
+    SDValue StartVal = getValue(Start);
+    Value *End = I.getArgOperand(1);
+    SDValue EndVal = getValue(End);
```
Avoid the temporary Value variable? StartVal = getValue(getArgOperand())

arsenm reviewed (approved):
Thanks!

rofirrim committed:
Address reviewer comment
